config.go

```diff
@@ -18,18 +70,79 @@ type Config struct {
 	InvalidUsernameMessage string   `json:"invalid-username-message"`
 }
 
+type ProxyPolicyConfig struct {
+	AllowedCIDRs []netip.Prefix
+	AllowedHosts []string
+}
+
 type UsernamePolicyConfig struct {
-	InvalidUsername        []string `json:"invalid-username"`
-	InvalidUsernameMessage string   `json:"invalid-username-message"`
+	InvalidUsernames       []string
+	InvalidUsernameMessage string
 }
 
 type PasswordPolicyConfig struct {
-	AllUsernameNoPassword bool     `json:"all-username-nopassword"`
-	UsernameNoPassword    []string `json:"username-nopassword"`
+	AllUsernameNoPassword bool
+	UsernamesNoPassword   []string
 }
 
-type RecoveryConfig struct {
-	Server   string   `json:"recovery-server"`
-	Username []string `json:"recovery-username"`
-	Token    string   `json:"token"`
+func convertLegacyConfig(config LegacyConfig) Config {
+	if config.RecoveryToken == "" {
+		config.RecoveryToken = config.Token
+	}
+	hostKeys := make([]SSHKeyConfig, 0, len(config.HostKeys))
+	for _, path := range config.HostKeys {
+		hostKeys = append(hostKeys, SSHKeyConfig{Path: path})
+	}
+	return Config{
+		Address: config.Address,
+		SSH: SSHConfig{
+			Banner:   config.Banner,
+			HostKeys: hostKeys,
+		},
+		Auth: AuthConfig{
+			Endpoint:               config.API,
+			Token:                  config.Token,
+			InvalidUsernames:       config.InvalidUsername,
+			InvalidUsernameMessage: config.InvalidUsernameMessage,
+			AllUsernameNoPassword:  config.AllUsernameNoPassword,
+			UsernamesNoPassword:    config.UsernameNoPassword,
+		},
+		Logger: LoggerConfig{
+			Enabled:  config.Logger != "",
+			Endpoint: fmt.Sprintf("udp://%s", config.Logger),
+		},
+		ProxyProtocol: ProxyProtocolConfig{
+			Enabled:  len(config.ProxyCIDRs) > 0,
+			Networks: config.ProxyCIDRs,
+		},
+		Recovery: RecoveryConfig{
+			Address:   config.RecoveryServer,
+			Usernames: config.RecoveryUsername,
+			Token:     config.RecoveryToken,
+		},
+	}
+}
+
+func convertProxyPolicyConfig(config ProxyProtocolConfig) (*ProxyPolicyConfig, error) {
```

Comment:
I'm not sure what's the point of returning a pointer here. Everywhere else that returns a config returns the `struct` directly, not through a pointer.

Reply:
Part of the reason is stated below. The other point is that we can avoid allocation on failed cases. I believe it's a convention to return `nil` for failable functions?